### ISSUE TITLE
Fx-9860 GIII

### ISSUE THREAD

**Issue by PhantomOverrideAlpha** [KBD2/terrario#1]
It just keeps reseting mycsio fx-9860 GIII could you make a compatible version for my calculator or give me a file that is compatiable

**Comment by PhantomOverrideAlpha** [comment]
You can email me @ Samuel.es.johnson@gmail.com

**Comment by KBD2** [comment]
Hi, thanks for notifying me. 
The problem is that Terrario uses 250kB of RAM that is available on the GII model but not the GIII. This can be fixed by instead using 80kB of the GIII heap, which means the world is a bit smaller.
I've attached a build that uses the heap. I don't have a GIII, so if you could confirm the build works that'd be great. If it does work, I'll look into adding a build for the GIII in each future release.
[terrario_v0.5.0-indev-GIII.zip](https://github.com/KBD2/terrario/files/5167416/terrario_v0.5.0-indev-GIII.zip)

**Comment by PhantomOverrideAlpha** [comment]
Ok so it kind of works it starts loading for about 5 seconds and then it resets the Calculator  i am not sure why that is

**Comment by KBD2** [comment]
Strange, that build works on both my GII and another tester's GIII.

**Comment by PhantomOverrideAlpha** [comment]
What is The OS version of the other testers Calculator

**Comment by PhantomOverrideAlpha** [comment]
I also have A Casio fx-9750 GIII if you have a way to make it compatible with that since the Casio fx-9750 GIII can and will accept addins in the SH4A format

**Comment by PhantomOverrideAlpha** [comment]
I tested it on my Casio fx 9860 GII and it worked but not on my Casio Fx 9860 GIII and Casio Fx-9750GIII

**Comment by lephe** [comment]
The build from https://github.com/KBD2/terrario/issues/1#issuecomment-686358607 crashes on my Graph 35+E II (French fx-9860G III) too, with an address error at 801ebbac, which is really close to the `malloc()` code on my OS version (3.10).

**Comment by PhantomOverrideAlpha** [comment]
Ok so how do we fix that and is there any chance that we could ake it work on a Casio Fx-9750 GIII

**Comment by KBD2** [comment]
Try this build, I identified a possible cause.
[Terrario.zip](https://github.com/KBD2/terrario/files/5178690/Terrario.zip)

**Comment by PhantomOverrideAlpha** [comment]
Ok it giveing me a error. is the malloc command non compatiable

**Comment by PhantomOverrideAlpha** [comment]
is there another command you can use instead of Malloc and how  about trying to save the game to storage memory (Rom)

**Comment by KBD2** [comment]
Which error is it giving you?
The game is saved to storage memory, but when you're playing it has to be loaded to RAM for performance reasons.

**Comment by PhantomOverrideAlpha** [comment]
801ebbac same as the other tester

**Comment by KBD2** [comment]
Can you try this one? It has all the storage memory calls stripped out, just in case they're what's causing it.
[Terrario.zip](https://github.com/KBD2/terrario/files/5178732/Terrario.zip)

**Comment by PhantomOverrideAlpha** [comment]
thats not it because it just back to reseting it do you have a reboot command in the code that is causing it?

**Comment by KBD2** [comment]
Nope, no reboots that could be reached just by starting the game.
Without a calculator to test on, I can't really do anything to fix all the errors and reboots. The errors stopping without `BFile` calls mean the game isn't really playable anyway without the ability to save the game.
This will probably have to wait until the gint runtime is fully compatible with the fx9750GIII.

**Comment by PhantomOverrideAlpha** [comment]
so is there another way to save the game

**Comment by KBD2** [comment]
No, the `BFile` calls are the only way to interface with the storage memory.

**Comment by PhantomOverrideAlpha** [comment]
what if we compile another addin to create a save file to and then run the terraio file

**Comment by KBD2** [comment]
Again, this would require that I have an fx9750GIII. I'm very handicapped for making this game work on the GIII, especially without much documentation on the exact differences between the two models that I could use to make the game compatible.

**Comment by PhantomOverrideAlpha** [comment]
there is 1 that i know of it is bazzoka here is the file maybe we could save th worlds inside of the g1a file https://www.planet-casio.com/Fr/programmes/programme865-last-bazzzooka-bebe-vador-a7.html

**Comment by PhantomOverrideAlpha** [comment]
that file i have on my calculator and it saves

**Comment by KBD2** [comment]
That game is much simpler and doesn't use the gint runtime, which is why it's a lot more likely to be compatible.

**Comment by PhantomOverrideAlpha** [comment]
yeah but i also have calc city and it a little more complex here is the file 

[Archive.zip](https://github.com/KBD2/terrario/files/5178799/Archive.zip)

**Comment by PhantomOverrideAlpha** [comment]
it save as a dat file instead of in game which could be a possiblity maybe with terraio

**Comment by KBD2** [comment]
Terrario already uses .dat files to save the world, and Calc City definitely loads that save file into RAM when you're playing the game. I can't do anything more than try to use `malloc`, because I can't actually test the game myself.

**Comment by PhantomOverrideAlpha** [comment]
can you give me a test dat file for a litle theroy i have

**Comment by PhantomOverrideAlpha** [comment]
i also am more than willing to help you make it compatible since i do have both calculators

**Comment by KBD2** [comment]
Here's a save game and addin build that would in theory work with the GIII.
[game.zip](https://github.com/KBD2/terrario/files/5178816/game.zip)
Fixing all the crashes over a GitHub issue or email (and we don't even know if they're all fixable) isn't really feasible, since it would involve dozens of builds and debug displays to fix things.

**Comment by PhantomOverrideAlpha** [comment]
did you take out the command that makes the save files?

**Comment by KBD2** [comment]
Not in that build, because you need it to load the game save.

**Comment by PhantomOverrideAlpha** [comment]
ok i was just checking

**Comment by PhantomOverrideAlpha** [comment]
ok so now it just keeps loading so we are getting somewhere

**Comment by PhantomOverrideAlpha** [comment]
how much ramcmemory does the game take up

**Comment by KBD2** [comment]
That build uses 30kB of RAM for the world save, and probably around 35-40kB at peak to store everything it uses the heap for.

**Comment by PhantomOverrideAlpha** [comment]
is there a way to transfer the files not being used to the storage memory because i think that is our problem

**Comment by KBD2** [comment]
There aren't any 'files' being used, it's all buffers and structs. 30kB is well within what the heap can hold, that's not the issue.

**Comment by PhantomOverrideAlpha** [comment]
ok hold on i have a idea

**Comment by PhantomOverrideAlpha** [comment]
it did not work
i tried resting my calc and running the game

**Comment by PhantomOverrideAlpha** [comment]
i am going to reserach and find out more about the diffrencs and get back to you tommorw ok

**Comment by PhantomOverrideAlpha** [comment]
did you examine the code and try and use the same sequence as calc city

**Comment by PhantomOverrideAlpha** [comment]
saving sequence

**Comment by KBD2** [comment]
I don't have the source code for that game. The internal structure of the save files are very different, Calc City's save sequence wouldn't work for Terrario.

**Comment by PhantomOverrideAlpha** [comment]
ok well we can work more on th\is once  have had a chance to compare the code also here ftune for SH4A 
[Ftune.zip](https://github.com/KBD2/terrario/files/5178880/Ftune.zip)

**Comment by PhantomOverrideAlpha** [comment]
if that helps you with diffrent commands for sh3 to SH4A

**Comment by KBD2** [comment]
The 9860GII and 9750GIII both use SH4A - in fact, Terrario isn't supported on SH3 calculators.

**Comment by PhantomOverrideAlpha** [comment]
i ment to say SH4 do you see any major code differences

**Comment by KBD2** [comment]
There is no 'SH4', only SH4A.

**Comment by PhantomOverrideAlpha** [comment]
though power graphic 2 ran on sh4

**Comment by KBD2** [comment]
No, SH4A.

**Comment by PhantomOverrideAlpha** [comment]
do you see any diffrence from 3.10(GIII) OS to 2.09 os (GII)

**Comment by KBD2** [comment]
https://www.planet-casio.com/Fr/forums/topic15699-1-test-de-la-graph-35e-ii-partie-1-applications.html

**Comment by PhantomOverrideAlpha** [comment]
is the problem the amount of pixels?

**Comment by KBD2** [comment]
No, I don't know what the problem is. It's likely to do with `malloc` but I have no idea why.

**Comment by PhantomOverrideAlpha** [comment]
![Screenshot 2020-09-05 at 7 57 56 PM](https://user-images.githubusercontent.com/70381526/92315332-46d80200-efb2-11ea-82eb-e204c2c3c8ec.png)
look

**Comment by PhantomOverrideAlpha** [comment]
is there a replacement command for malloc

**Comment by KBD2** [comment]
That's a different mofel calculator...

**Comment by PhantomOverrideAlpha** [comment]
No that is the techinal name in french for the casio fx 9750 GII

**Comment by PhantomOverrideAlpha** [comment]
cant you use the `free` command?

**Comment by KBD2** [comment]
I know the Graph 35+E II is the fx9750GIII equivalent. It has a display size of 128x64 pixels, the same as the fx9860GII.

**Comment by PhantomOverrideAlpha** [comment]
sorry but in all reality there has to be a longer and yes but i thought you could use that as a replacement or aleast that was what i was told is that wrong

**Comment by KBD2** [comment]
https://en.cppreference.com/w/c/memory/malloc Here is how the malloc command is used.

**Comment by PhantomOverrideAlpha** [comment]
ok so free still had something to do with malloc it just wsnt what i was told

**Comment by PhantomOverrideAlpha** [comment]
so what command does calc city use

**Comment by KBD2** [comment]
Calc City most likely allocates memory for the map, loads the map from file into the memory, then does the opposite when saving the game.

**Comment by PhantomOverrideAlpha** [comment]
is there a way we can do that

**Comment by KBD2** [comment]
That's what Terrario already does.

**Comment by PhantomOverrideAlpha** [comment]
so what could possibly be the problem

**Comment by KBD2** [comment]
I don't know.

**Comment by PhantomOverrideAlpha** [comment]
am about to compare the two codes real quick

**Comment by PhantomOverrideAlpha** [comment]
how does the terrario code start out

**Comment by PhantomOverrideAlpha** [comment]
how about you try this just to test it switch all the saving over to the storage memory and run  everything off of storage memory

**Comment by KBD2** [comment]
We've already established the `BFile` functions don't work. Additionally, as already explained I can't stream everything out of the storage memory because it's very slow.

**Comment by PhantomOverrideAlpha** [comment]
ok i am going to do some research on this and post a there on the universal casio forum to see if anybody can find a way to fix this probelm if you dont mind .do you mind?

**Comment by KBD2** [comment]
Go ahead, I'm out of ideas. Remember that Terrario uses something called gint, which makes it different from most addins.

**Comment by PhantomOverrideAlpha** [comment]
ok here is the link to the post https://community.casiocalc.org/topic/8021-help-making-terrario-compatiable/

**Comment by lephe** [comment]
>  This will probably have to wait until the gint runtime is fully compatible with the fx9750GIII.

Quick reminders. The fx-9860G III, fx-9750G III and Graph 35+E II all have the same ROM size, same applications, and they all support Python. Supporting Python is the key to the heap, because Python is the biggest heap user of them all. Try to run [Critor's Python heap memory estimator](https://tiplanet.org/forum/viewtopic.php?t=22913&p=244834) first to determine the amount of heap memory available to the user. The only documented differences between the fx-9750G III and the fx-9860G III are the unsupported Python drawing modules and a stricter examination mode.

Secondly, the Graph 35+E II has a different display than the previous GII series. The difference is not in pixel size, nor in surface, it's in *protocol*. You can see the differences in [gint's T6K11 driver](https://gitea.planet-casio.com/Lephenixnoir/gint/src/branch/compat/src/t6k11/t6k11.c#L16). A good way to tell if the screen is the problem is as follows: a crash/reboot will immediately restart the calculator without you needing to do anything, but a screen protocol failure will blank out the screen without rebooting, forcing you to press the RESET button to go back to the main menu.

And finally, I only have a Graph 35+E II so I have never tested my applications on international GIII models. If you suspect the problem is with gint itself, please build and run gint's testing application [gintctl](https://gitea.planet-casio.com/Lephenixnoir/gintctl) and try to spot any problem here. I can walk you through it. I would appreciate bug reports in [gint's repository](https://gitea.planet-casio.com/Lephenixnoir/gint). The SH3/SH4 difference is most likely irrelevant since all these machines are SH4 and gint is binary-compatible with both anyway.

As KBD2 mentioned, there is practically no way that `malloc()` is completely broken or that using it the way Terrario does is incorrect. I'll try the builds in this issue on my Graph 35+E II later and report back.

**Comment by PhantomOverrideAlpha** [comment]
OK thank you

**Comment by PhantomOverrideAlpha** [comment]
anybody found out anything or anything i can do to help

**Comment by PhantomOverrideAlpha** [comment]
So where are we at on compiling a version that is compatible with the casio fx-9750 GIII and what did your tests do.

**Comment by PhantomOverrideAlpha** [comment]
found a link that might help but i can't read french so you will have to tell me if it helps  https://www.planet-casio.com/Fr/forums/topic13572-1-gint-un-noyau-pour-developper-des-add-ins.html

**Comment by KBD2** [comment]
lephe is the developer of gint.

**Comment by PhantomOverrideAlpha** [comment]
oh i did not  know that he was the same person on planet casio

**Comment by PhantomOverrideAlpha** [comment]
I have summoned more help but i am waitng for them to get back with me one specializes in addins and assemby and another has made some addins for the Casio fx-9860 GIII and the Casio fx-9750GIII

**Comment by lephe** [comment]
Here's a recap of all the builds on the Graph 35+E II OS 3.10.

* https://github.com/KBD2/terrario/issues/1#issuecomment-686358607 (initial 30k heap): fails, address error at 801ebbac around `malloc()`
* https://github.com/KBD2/terrario/issues/1#issuecomment-687666287 (previous build amended): fails, same error
* https://github.com/KBD2/terrario/issues/1#issuecomment-687669874 (no BFile): **fails, same error**
* https://github.com/KBD2/terrario/issues/1#issuecomment-687673318 : fails, same error

So not a lot to blame on BFile I suppose, which is for the better because debugging gint's BFile and return-to-menu is always 10x harder than anything else.

@KBD2 I'd like to build and do some debugging myself, so I'll have to build a new toolchain with newlib support (I've had significant issues with it in the past so I dropped it at some point). In the meantime...

* In [`testRAM()`](https://github.com/KBD2/terrario/blob/master/src/main.c#L47), you should restore `*RAMTestAddress` *at all costs*, you can't write some data and leave if it works. Here is for instance how gintctl does this test (to date, successfully) : [`gintctl/src/gint/ram.c#L8`](https://gitea.planet-casio.com/Lephenixnoir/gintctl/src/branch/master/src/gint/ram.c#L8).
* Did you properly check that `world.tileData` is not NULL? As I mentioned before, the NULL pointer is mapped to a cached view of the boot sector, so you can in fact write there. It's unlikely to cause significant problems, but you never know.
* (On non-GIII builds you zero out `world.tileData` before even checking that `RAM_START` the area is writable, which will definitely wipe the initial RAM at `88000000` on 256k-RAM models. I think all of these are SH3 so Terrario is not compatible, but you can fix this easily by calling `testRAM()` before using the memory, or use `gint[HWRAM]` from [`<gint/hardware.h>`](https://gitea.planet-casio.com/Lephenixnoir/gint/src/branch/master/include/gint/hardware.h) which tells you how much RAM is available.)

**Comment by KBD2** [comment]
I've already implemented `malloc()` `NULL` checks, but I'll fix the other issues.

**Comment by KBD2** [comment]
Newest commit should have fixed the issues you noted.

**Comment by PhantomOverrideAlpha** [comment]
ok i will test it right now

**Comment by PhantomOverrideAlpha** [comment]
are we going to compile a test file that i can test on my 9750 GIII and or 9860GIII

**Comment by KBD2** [comment]
Here's a build from the latest commit, though I doubt anything will have changed.
[Terrario.zip](https://github.com/KBD2/terrario/files/5187933/Terrario.zip)

**Comment by lephe** [comment]
Same error indeed. This is an interesting situation.

**Comment by PhantomOverrideAlpha** [comment]
yeah i got the same error too

**Comment by PhantomOverrideAlpha** [comment]
could we be looking at this the wrong way

**Comment by PhantomOverrideAlpha** [comment]
is there a emulator for the casio fx-9750GIII

**Comment by PhantomOverrideAlpha** [comment]
I was reading and i came across a article on Planete Casio that said if the addin uses any getkey function otherthan getkey or getkey wait syscall will not work

**Comment by PhantomOverrideAlpha** [comment]
here is the link to the article https://www.planet-casio.com/Fr/forums/topic11632--.html scroll to bottom and read comments

**Comment by KBD2** [comment]
As previously stated, Terrario is written for SH4 calculators and doesn't need the compatibility tool (gint has its own keyboard interface so avoids the `getkey` problem).

**Comment by PhantomOverrideAlpha** [comment]
Ok so what are we going to do are we leaving the issue open so we can find out more or should i open up a thread on the universal casio forum to discuss it more there

**Comment by lephe** [comment]
I had to spend some time building a new toolchain with newlib, because Memallox's port does not actually support `m4-nofpu` as a multilib target (the newlib floating-point code wasn't extended with that in mind), and that target is included in my normal setup. It took me some time to troubleshoot the details... then rebuild the toolchain, reinstall gint and the libraries.

I was able to build a basic version of Terrario that worked as expected on the fx-9860G II and failed as expected on the Graph 35+E II. Next comes the actual debugging (finally).

**Comment by PhantomOverrideAlpha** [comment]
Ok how can I help with that

**Comment by PhantomOverrideAlpha** [comment]
I can help do some debugging

**Comment by KBD2** [comment]
Is there any detriment to using m3 instead of m4-nofpu?

**Comment by lephe** [comment]
I tracked down the crash to gint's gray engine constructor which allocates the gray VRAMs (the only place in gint where the heap is used). The first of three calls to `malloc(1024)` causes the crash. This is clearly not a problem of size. Additionally, gintctl has the exact same initialization and works just fine. I don't know yet what difference between the two applications causes this problem.

> Ok how can I help with that

Please try and run solutions on your international GIII models when tentative fixes are made :smiley: 

> Is there any detriment to using m3 instead of m4-nofpu?

There are slight advantages to using `m4-nofpu`, most of which are about recognizing SH4-only opcodes such as `synco` and `movua.l` in the assembler. Nothing that outweighs having newlib I guess.

**Comment by KBD2** [comment]
If there's no substantial performance decrease then I definitely prefer having newlib, as the math library is invaluable for world generation.

**Comment by PhantomOverrideAlpha** [comment]
Ok i have no probelm testing

**Comment by PhantomOverrideAlpha** [comment]
I just made a discovery here Version 0.0 works on my Casio fx9750 GIII Version 1.0 works on my calculator version 2.0 and up do not work and if Lephe can confirm my results we can figure out whats diffrent

**Comment by KBD2** [comment]
0.0.0 and 0.1.0 use the heap instead of the spare RAM as they don't save the world, and not many things are implemented at all in those versions.

**Comment by PhantomOverrideAlpha** [comment]
ok so is there any way to fix it

**Comment by KBD2** [comment]
That's the entire purpose of this issue. I'm pretty sure lephe has identified the source of the main crash.

**Comment by PhantomOverrideAlpha** [comment]
I know that i wonder what is main crash caused by

**Comment by lephe** [comment]
It is possible that the memory layout of the add-in overwrites some internal data when too large. Terrario is much larger than gintctl. Honestly the only code that executes before the constructors is (1) copying the add-in to RAM, (2) setting up the kernel, (3) loading the drivers, so there are not many possible culprits.

Next thing I'll check is whether a large gintctl has the same problem, and whether there are influential drivers. Sorry I am not able to progress very fast, I wish I could spend more time here.

**Comment by PhantomOverrideAlpha** [comment]
> It is possible that the memory layout of the add-in overwrites some internal data when too large. Terrario is much larger than gintctl. Honestly the only code that executes before the constructors is (1) copying the add-in to RAM, (2) setting up the kernel, (3) loading the drivers, so there are not many possible culprits.
> 
> Next thing I'll check is whether a large gintctl has the same problem, and whether there are influential drivers. Sorry I am not able to progress very fast, I wish I could spend more time here.

Is there anything i can test for you

**Comment by PhantomOverrideAlpha** [comment]
anything to test yet

**Comment by lephe** [comment]
Nothing yet. Please don't spam messages here, I'll post as soon as I make progress, you'll be notified in no time. Additional messages just make it harder to browse the comments.

**Comment by PhantomOverrideAlpha** [comment]
OK so I kinda figured out a list of addins that work on the Casio fx9750GIII if it helps any let me know most of them are very simple BTW

**Comment by PhantomOverrideAlpha** [comment]
Ok soI just installed Insight And it says my Heap Max is 88050400 what does that mean

**Comment by lephe** [comment]
I didn't know Insight provided this information. This is the size of the heap - 88 kB, which is pretty much what Critor's estimator had determined previously. This is the size that the heap has been increased to in the GIII models in order to support moderately complex Python scripts.

**Comment by lephe** [comment]
While investigating a different Graph 35+E II-only System ERROR I discovered that gint's VBR address is misplaced on this platform. This issue is sometimes pretty severe in terms of how it can interfere with an add-in, and I am surprised it did not come up during testing for release 2.1. By loading data to an incorrect address, gint currently overrides some OS data, which causes all sorts of unexpected behaviors, including that System ERROR I was debugging.

I have confirmed that this VBR misplacement is *also* causing the heap corruption problem. Please find attached a Terrario build compiled with a gint version that hardcodes the proper location on a Graph 35+E II. I would like for it to be tested on a GIII. Additionally, please start the gintctl application on the GIII, navigate to the second tab (GINT), down to "TLB Management". Here, inspect the TLB on the third tab and tell me what physical address `08100000` is mapped to (usually way down the list). Older platforms used `0801c000` but the Graph 35+E II uses `08048000`, which I had never realized. (This is in the second half of the RAM. One more reason why using that half on a GIII is by all means unsafe.)

A proper solution to this problem requires me to relocate data on some platforms. I will publish a gint fix as soon as I can find and thoroughly test a suitable method.

[terrario-gintctl.zip](https://github.com/KBD2/terrario/files/5213163/terrario-gintctl.zip)

**Comment by PhantomOverrideAlpha** [comment]
Ok so the game works And saves as to theother addin it gives me this 
![PIXECT-20200912173847](https://user-images.githubusercontent.com/70381526/93005461-172f7980-f51f-11ea-8518-9745e8a9df39.jpg)

**Comment by PhantomOverrideAlpha** [comment]
At times The addin gives me thi
![PIXECT-20200912182947](https://user-images.githubusercontent.com/70381526/93006088-fa4a7480-f525-11ea-957c-2ef8095bc0aa.jpg)
s screen

**Comment by KBD2** [comment]
This is some great progress, I would never have been able to diagnose that issue.

**Comment by PhantomOverrideAlpha** [comment]
i can play the game and it will save but at times it will display a error but it is playable

**Comment by KBD2** [comment]
Can you figure out if there is any specific action that results in a crash?

**Comment by PhantomOverrideAlpha** [comment]
I will test and see but the error happes when mostly happens when you go to the About part and then go to new world i gives a error whch looks like this`Execption! (SysERROR)
Read address error
 PC :00000001
TEA :00000001
TRA :00000000
SGR:ffffffff`

**Comment by PhantomOverrideAlpha** [comment]
yes it is when you go from the about page then go to the the new game

**Comment by KBD2** [comment]
I can't reproduce that on my 9860GII.

**Comment by PhantomOverrideAlpha** [comment]
I can make a video showing you it

**Comment by PhantomOverrideAlpha** [comment]
I think found a emulator for the casio 960GIII and casio 9750 GIII

**Comment by lephe** [comment]
Can reproduce the About bug. I'm currently putting up a solution for the VBR placement problem, will update you when this is available in the gint tree.

> I think found a emulator for the casio 960GIII and casio 9750 GIII

The official Graph 35+E II emulator is also a good start.

**Comment by PhantomOverrideAlpha** [comment]
That is the one I am talking about 
https://www.planet-casio.com/Fr/logiciels/voir_un_logiciel_casio.php?cat=7

**Comment by hejsotnoss** [comment]
Hi I see you reached out to me on casiocalc.org. I am not experienced much with gint since I usually write my addins in assembly. As lephe said the VBR address seems to be misplaced for the Fx-9860GIII, so I think waiting for them to release a patch would be the solution. I also saw that you mentioned syscalls which don't seem to be an issue since most of the os versions I've looked into seem to have backwards compatibility support. Thank you for reaching out to me.

**Comment by PhantomOverrideAlpha** [comment]
Are we closing this issue or is there still more work to be done

**Comment by lephe** [comment]
I am working on fixing the original VBR problem right now. The About bug will also need some investigation (which can be done here or on another issue).

**Comment by PhantomOverrideAlpha** [comment]
we can get it do here is there a program in which I can figure out what  going wrong with the About bug

**Comment by KBD2** [comment]
I've identified a possible cause (if the About menu is left early via [EXIT] a timer is not stopped), I'll work on a fix when I can.

**Comment by lephe** [comment]
The loading problem has now been fixed in gint, please see [commit `e5abe03`](https://gitea.planet-casio.com/Lephenixnoir/gint/commit/e5abe03b89f544e4cf7f5fb5aa93a6d3d60ca73e) (also tagged 2.1.1). This is a large change to the kernel to detect the user RAM address at runtime and load code here. I am confident that this will improve compatibility for future models and OS versions.

I have checked that Terrario builds and runs correctly as in https://github.com/KBD2/terrario/issues/1#issuecomment-691535869, but an additional check would be welcome before this issue can be closed. :slightly_smiling_face:

**Comment by KBD2** [comment]
Awesome, I'll update gint and make test builds when possible.

**Comment by PhantomOverrideAlpha** [comment]
I will test it now

**Comment by KBD2** [comment]
I've updated gint and created 3 builds I need tested, to see how much heap memory I can use:

30kB heap usage - [Terrario_30kB.zip](https://github.com/KBD2/terrario/files/5243852/Terrario_30kB.zip)

60kB heap usage - [Terrario_60kB.zip](https://github.com/KBD2/terrario/files/5243855/Terrario_60kB.zip)

75kB heap usage - [Terrario_75kB.zip](https://github.com/KBD2/terrario/files/5243856/Terrario_75kB.zip)

Please test these and tell me if any don't run. The first one definitely should, as it's the one Lephe would have been testing with.

**Comment by lephe** [comment]
The first one works on my Graph 35+E II, unfortunately the two others produce memory errors. Maybe the heap is fragmented and allocating in several blocks would allow for more memory to be used?

On a different level, you may not need to reboot when the error occurs. Some options include returning from `main()` which (by default) quits the add-in in a way that the user can't come back, returning from `main()` after `gint_setrestart(1)`, which restarts the add-in if the user comes back, or entering a loop of `gint_osmenu()` to continually send the user back.

**Comment by KBD2** [comment]
I've implemented the forced main menu after a `NULL` `malloc()` and changed the GIII world size to 400x100 (40kB heap usage), which is the maximum size I know will run (works on my GII). I'll start including this build in the next releases of Terrario.

Splitting the data would require a large rework (and quite a bit of overhead) of the way functions access the world data, so while I have it on my to-do list it isn't a priority.

As the builds are now proven to work on the 35+E II/GIII, I think we can finally consider this issue solved.

**Comment by KBD2** [comment]
If there are no last comments, I'll close the issue in a bit.

**Comment by Scratchdragon** [comment]
Seems this issue is preventing me from playing version 0.10.0 on my GIII, I was thinking of making a fork to resolve it but was just wondering if there are any plans to create a specific GIII build in the future to resolve this?

**Comment by KBD2** [comment]
Unfortunately I don't have a GIII so I wouldn't be able to find the source of the issue myself. I've stopped active development for the foreseeable future as I'm preoccupied with work and university, but feel free to fork and I'd be happy to merge it if you can find a fix.

**Comment by lephe** [comment]
I've successfully played a number of versions on a Graph 35+E II (fx-9860G III) and I thought this bug was resolved for good. The G-III series should be officially supported as far as I understand.

The fix for the issue here was integrated in v0.5.1-indev if we trust the release dates. Could you possibly check whether versions between v0.10.0 and v0.5.1-indev work? This would help both with determining whether your problem is indeed linked to this issue, and if not what version introduced it.

**Comment by Scratchdragon** [comment]
I went through all of the releases from 0.5.1 to 0.9.0, all of the versions ran fine with no error except 0.6.0 which gave the error : "Mem alloc returned NULL! Please report this bug.". I also went back to confirm the error I get with 0.10.0 is the same as in this error and can confirm it is the same.

**Comment by lephe** [comment]
Thank you for testing extensively. So that seems to be simply the generic out of memory problem. @KBD2 were there any memory-heavy additions in 0.10.0?

Maybe we can find more memory but I believe the map is already stored in PRAM, so we're getting somewhat tight on that front.

**Comment by KBD2** [comment]
The only change I can see affecting memory usage is `MAX_ENTITIES` being changed from 5 to 12, but the entity structure isn't that large that I can see it being the cause of the issue here.
The 5 places heap memory is allocated in 0.10.0 are:
- Any new chat message structs
- The actual text for said message
- World entities
- Explosion particles
- Chests
You could try fiddling around with the values for these to see if it has just hit the heap limit for the GIII.

**Comment by RaidTheWeb** [comment]
On my CASIO fx-9860GIII with the latest release I encounter an issue that always happens after generating a new world.

After moving around for a few seconds the calculator encounters an exception as follows:

TLB miss read
PC:   0031ca24
TEA: 0033b046
TRA: 00000000
SGR: 8804fe64

Upon saving before such a crash occurs this doesn't happen when you load back in, instead there is just no terrain and you fall infinitely until you die.

Just thought I'd add my problems to the list :)

**Comment by KBD2** [comment]
Check #7 - I think it may be an issue with the world generation code, as I haven't really looked at it since the last PR.

**Comment by RaidTheWeb** [comment]
That'd likely be the case, I'd try fix it myself if I could actually get gint and the cross compiler working. :roll_eyes:

**Comment by lephe** [comment]
> That'd likely be the case, I'd try fix it myself if I could actually get gint and the cross compiler working. :roll_eyes:

Sorry if you're having problems with that. I can help you troubleshoot it, feel free to post as a guest on Planète Casio, for instance [here](https://www.planet-casio.com/Fr/forums/topic13164-last-fxsdk-un-sdk-alternatif-pour-ecrire-des-add-ins.html).

**Comment by RaidTheWeb** [comment]
> > That'd likely be the case, I'd try fix it myself if I could actually get gint and the cross compiler working. roll_eyes
> 
> Sorry if you're having problems with that. I can help you troubleshoot it, feel free to post as a guest on Planète Casio, for instance [here](https://www.planet-casio.com/Fr/forums/topic13164-last-fxsdk-un-sdk-alternatif-pour-ecrire-des-add-ins.html).

It's all good, I just figured it out myself after some thought.

**Comment by RaidTheWeb** [comment]
what I can confirm though, is that after building the latest commits I haven't encountered any more issues relating to the aforementioned one. The only issue that remains is that loading a world does not load in any terrain and rather drops you infinitely into the void. :D

**Comment by KBD2** [comment]
If you're using a GIII the issue is likely with PRAM access (the world is stored there on that model) or writing files, though I hadn't heard of any issues before now. It's possible something isn't being initialised or accessed correctly - @lephe is a lot more knowledgeable on this topic and might be of more help.

**Comment by lephe** [comment]
Just to be sure, is there no risk of confusing "no terrain loaded" with "spawned high up in the sky"? I think you'd see the terrain after dying in the second case, but I'd rather confirm it.

The PRAM should normally not be a problem. Any issue with writing the files could be diagnosed by manually looking at a  backup of the world. @RaidTheWeb could you share your chunk files with us or check them in a hex editor?

I don't know of any special condition occurring with PRAM; the only real limitation while writing files is to not write from ROM. I'm not ruling DMA oddities entirely, but it sounds like if it didn't work if would just *never* work. This could be tested around by loading chunks to RAM before copying them to PRAM, and the other way around for saves.

By the way @KBD2, gint now integrates filesystem access with standard functions and while implementing it I noticed that overwriting existing files is much faster than creating new files on the G-III; not removing existing files is probably a low-hanging fruit for good performance improvements. It doesn't matter if the new chunk data is longer since the file would expand, and I assume it wouldn't matter to your code if it's smaller either. (This only works on the Fugue FS, which you can check for in `gint[HWFS]`.)

**Comment by RaidTheWeb** [comment]
This seems to be consistent in the official emulator as well as on my actual calculator, and no, the terrain has physically not loaded you fall infinitely until you fall off the screen and die.

**Comment by tt-thoma** [comment]
I don't understand... [This page ](https://www.planet-casio.com/Fr/infos/g35+e2) referencing the Graph 35+E II stats says that the calculator has a 60KB memory. Doesn't that means the calculator should crash when there is a heap size of 60KB or more ? The TLB error seems to point out this exact problem.
I think you're searching too far

**Comment by lephe** [comment]
The page you're linking isn't talking about the actual hardware specs; it's about normal, math class usage of the calculator. Specifically 60 kB is the size of the so-called _"main memory"_, which is the storage for built-in app data related to functions, statistics, sequences, matrices, and the like.

The G-III and its predecessors still have a 512-kB RAM chip, out of which 60 kB are reserved for the main memory, and _another_ 48 kB (G-II) / 90 kB (G-III) are reserved for the heap.

I'm not sure which comment you're responding to though - the latest TLB error issue mentioned in the thread [appears to be solved already](https://github.com/KBD2/terrario/issues/1#issuecomment-1034670233).

**Comment by tt-thoma** [comment]
It's not, I have the same issue. 
It's weird to see the exact same crashing values everywhere... I think it has still something to do with the TLB.
I documented myself a bit on Wikipedia, and it looks like when there is a TLB miss, the system should look up an address in another table. Could this be the problem or is this function already handled ?
```
Exception! (SysERROR)
TLB miss read
 PC :0031ca24
TEA :0033b046
TRA :00000000
SGR :8804fe64
The add-in crashed.
Please reset the calc
```

**Comment by lephe** [comment]
In modern computers (ie. what Wikipedia is telling you about), page tables can be pretty huge so they're stored in RAM and the TLB is a cache of recently-accessed pages. When the TLB is not hit, the real page table in RAM is traversed. However, on the calculator page tables are small and therefore they fit entirely within the TLB. Hence, a "TLB miss" is actually a page fault.

In the System ERROR screen, the value of PC tells you the address of the instruction which raised the exception (ie. the faulty memory access) and the value of TEA usually tells you which invalid address was involved. Both of them are in the `00300000` region, which is where the g3a file is mapped (for code and read-only data).

The program made a code or data access at `0033b046`, which is presumably past the end of the g3a file. This is a fairly weird occurrence. My bet is that the bug is not on the instruction at PC, but rather somewhere upstream, and it simply cascaded until that particular memory access got out of bounds.

If you have access to the build folder, check if you have a "map" file in it, or change the CMakeLists.txt to add the link flag `-Wl,-Map=map` then rebuild and look in `build-fx/map` to find which function is at `0031ca24` and if there is something at `0033b046`. Alternatively, you can disassemble the ELF and work your way up from here, if you're familiar with that stuff.

I don't have any cycles to look at this in more detail right now, sorry...

**Comment by tt-thoma** [comment]
I just have one question: My calculator is actually a graph 35+E, but with seemingly Graph 35+E2 specs (it only runs program made for the Graph 35+E2 and seems to have the same memory limitations). Is it normal ?

**Comment by KBD2** [comment]
Hey, unfortunately I still don't have the right model to bugfix that on - and I'd have to remake my development toolchain, it's been a couple years (and a new computer) since I last actively developed this project.

**Comment by tt-thoma** [comment]
@lephe Oh you're the one who coded the panic screen, it makes much more sense now

**Comment by lephe** [comment]
> I just have one question: My calculator is actually a graph 35+E, but with seemingly Graph 35+E2 specs (it only runs program made for the Graph 35+E2 and seems to have the same memory limitations). Is it normal ?

You sure it's not a 35+E II? The "II" is small and fairly easy to miss, it might have worn off too. For various technical reasons I'm very confident that there exists only pre-G-III machines with pre-G-III hardware and G-III machines with G-III hardware. Check your OS version; if it's 2.xx you have a G-II, if it's 3.xx you have a G-III. If in doubt, send a photo...

> Oh you're the one who coded the panic screen, it makes much more sense now

Yes I made the kernel/SDK, which is why I lurk around for debugging

**Comment by tt-thoma** [comment]
Yep I got the latest v3 version (03.40.2200).

Edit: the calculator is indeed a 35 E+2

**Comment by tt-thoma** [comment]
I was messing around with the official Micropython integration and tried an operation that would fill up the memory. Python said that it could fill up to 40708 bytes before raising a memory error, no matter what was already in the memory

**Comment by tt-thoma** [comment]
I found the problem!
Edit (clarification): It was caused by the clock addon you made. Deleting the clock removes the problem

**Comment by lephe** [comment]
Wait, sorry? I'd really like more details here, because that makes very little sense to me. You can only run one add-in at a time - if you're running Terrario, any clock add-in that you might have installed on the side is as good as if it didn't exist at all. Also I'm pretty sure I have written no add-ins that run "in the background", because these are annoying to work with. Finally, gint interacts with the system in lots of ways (through memory, filesystem, etc) so "couldn't be the system" is a very optimistic assumption...

**Comment by tt-thoma** [comment]
I don't really know, but you probably know how this [SH4 clock](https://www.planet-casio.com/Fr/programmes/programme2606-last-horloge-sh4-lephenixnoir-utilitaires-add-ins.html) was made, since you kinda... made it ?

**Comment by lephe** [comment]
Sure, but everything I know tells me it definitely has nothing to do with the bug. If you asked me randomly, I'd rather bet on non-determinism than this clock having an influence. Which is why I'm curious to see what observations led you to think this is the cause of the bug.

**Comment by tt-thoma** [comment]
I deleted the clock program and it just worked normally after... Do you have something like an emulator that can debug this ?

**Comment by lephe** [comment]
Ah, so maybe a filesystem effect that got resolved after removing a file... there is the official CASIO emulator, but there aren't any debugging-specific tools in it (it just runs stuff). Mostly I'm afraid that's the type of issues that's annoyingly difficult to reproduce. A clean reproduction method would be the most useful thing to have atm.

**Comment by tt-thoma** [comment]
I'll try reproducing the problem on my calculator

**Comment by tt-thoma** [comment]
Can confirm terrario not working with the clock
But it also crashes without the clock

**Comment by lephe** [comment]
Does optimizing the file system or otherwise adding, removing, changing files (if possible without re-saving the world) help?

**Comment by tt-thoma** [comment]
Nothing works anymore

**Comment by tt-thoma** [comment]
The only thing that might have made a difference was an optimization was turned off message after saving a world

**Comment by tt-thoma** [comment]
Optimizing does nothing neither

**Comment by Dodecahedron7777** [comment]
Hello, I also have a casio fx9750GIII (updated to os 3.7),
and I wanted to point out some specific error crashes.

On vs 10.0
jumping:
although it has crashed in the past everytime I jumped,
This time when I loaded it to my calculator it worked.
Going too far:

**TLB miss read
PC :0031ca24
TEA :0033b046
TRA :00000000
SGR :8804fe64**

if you press **OPTN**, the calculater restarts.

vs 9.0
if you press **OPTN**, the world dissapears.
This led me to thinking that when data is loaded to or from storage, the world dissapears.
And since saving and loading the world is all about storage, the world disapears when you select the load game option.

You may already have this data, but I wanted to present my ideas.

**Comment by Dodecahedron7777** [comment]
Also, if you don't own a fx9750GIII, why not try using an [enumerator](https://www.planet-casio.com/Fr/logiciels/voir_un_logiciel_casio.php?showid=144)?

**Comment by tt-thoma** [comment]
I think the problem comes from the saving system

**Comment by Dodecahedron7777** [comment]
Me too.
So, how do we fix it?

**Comment by tt-thoma** [comment]
Isn't the program trying to use some memory allocated to the system ? I've seen games like Wolfenstein unload the system menu before starting

**Comment by tt-thoma** [comment]
Also could you release a version with 30kb memory ? It could be a temporary fix

**Comment by lephe** [comment]
> Isn't the program trying to use some memory allocated to the system ? I've seen games like Wolfenstein unload the system menu before starting

There was such a bug before, but no longer. "Unloading the system menu" isn't a thing as far as I know.

> Also could you release a version with 30kb memory ? It could be a temporary fix

I don't think the game will even run on 30 kB.

Just to be clear, the "clean" way to solve this bug would be to delve back into the code and fix the likely-innocuous issue causing it. A first step towards that would be to have a way to consistently reproduce the bug. As long as it's random, it'll be difficult to check that it is even fixed when attempting changes.

**Comment by tt-thoma** [comment]
> There was such a bug before, but no longer. "Unloading the system menu" isn't a thing as far as I know.

I don't know what truly happens, but the game seems to have the calculator to restart entirely when exited. 
Also I'm not sure, but the problem might be that the system tries to use memory space allocated to the game. (that could explain the outside of app instruction error)

> Just to be clear, the "clean" way to solve this bug would be to delve back into the code and fix the likely-innocuous issue causing it. A first step towards that would be to have a way to consistently reproduce the bug. As long as it's random, it'll be difficult to check that it is even fixed when attempting changes.

I understand, but right now the game is straight unusable afaik

**Comment by Dodecahedron7777** [comment]
Does it only not work on the fx-9850G lll?
If it doesn't work, could we downgrade from fx-9850G III to fx-9750G III?

**Comment by jmerono-saez** [comment]
> I understand, but right now the game is straight unusable afaik

I mean, I still have it running kinda properly on my fx-9860 GII, only crashing every now and then. If anyone has any idea of where to start getting the bug to reproduce more consistently, I would love to try helping out (;

**Comment by tt-thoma** [comment]
> I mean, I still have it running kinda properly on my fx-9860 GII, only crashing every now and then. If anyone has any idea of where to start getting the bug to reproduce more consistently, I would love to try helping out (;

Oh on my side it's straight crashing when going too far left or right

**Comment by Joyful-Diamond** [comment]
Hey, I have two of the games installed ( the gint runtime and the other 30kb heap) on my Casio Fx9860GIII calculator. The storage memory has been filled (EXTREMELY likely that the cause was these programs) and both of them are kind of the same program? like if I use either then it's the same world.

Anyway, because of this I now have an input-output error, and I cannot delete any of these files. I have tried literally everything, save for using a debug menu which I do not know how to access. I have reset everything, connected to a PC, etc etc. Also, I deleted some files responsible for the tools for the haha, and then now I have a world where I cannot use any tools since I deleted the inventory data.

if there is any way this could be solved I would be very appreciative. especially just telling me what button combination is the debug menu.

**Comment by tt-thoma** [comment]
> Anyway, because of this I now have an input-output error, and I cannot delete any of these files. 

Sounds like a FS corruption issue

Try running chkdsk or fsck-ext on your calculator drive (plugged in your pc - this might not work cuz idk how the FS update works)

**Comment by Joyful-Diamond** [comment]
C:\Windows\System32>chkdsk d: /r
The type of the file system is FAT.
Windows is verifying files and folders...
94 percent completed.....
C:\Windows\System32>

:(

**Comment by lephe** [comment]
Maybe relevant: https://www.planet-casio.com/Fr/forums/topic16521-last-fx-9750giii-storage-memory-crash.html

For context, there's been a few people with such filesystem corruption issues using various add-ins or official apps. Very hard to deal with when it can be dealt with. I suspect it's an OS bug, but unclear. CASIO is aware but it's hard enough to replicate that they can't do much about it, I fear.

**Comment by tt-thoma** [comment]
> For context, there's been a few people with such filesystem corruption issues using various add-ins or official apps. Very hard to deal with when it can be dealt with. I suspect it's an OS bug, but unclear. CASIO is aware but it's hard enough to replicate that they can't do much about it, I fear.

It might just be a hardware issue, after a prolonged use it's just common for storage cells to fail

Also, can't they have a file system repair? AFAIK they have access to all of their hardware and could simulate failing blocks to build a program to remove them

**Comment by lephe** [comment]
> It might just be a hardware issue, after a prolonged use it's just common for storage cells to fail

Maybe, but it's strange that it only happens on this model specifically.

> Also, can't they have a file system repair? AFAIK they have access to all of their hardware and could simulate failing blocks to build a program to remove them

There are filesystem functions that can do that sort of thing; just cleaning up open descriptors that are left lying around, "optimization" (defragmentation), and the so-called "low-level format" are good targets to try and improve the situation. But finding ways to call them when you can't transfer any more files is difficult. C.Basic is useful with that. On some models the low-level format function is available in a debug menu. Tragically, the G-III is not one of them.

**Comment by caiji-xd** [comment]
I've found that version 0.5.1 doesn't have saving issues when running on fx-9860GIII, maybe this is a possible solution

**Comment by caiji-xd** [comment]
all versions before v0.10 and after v0.5.1 can run with stablilty on fx9860GIII (but can't save worlds correctly)

**Comment by Joyful-Diamond** [comment]
Yeah i can't fix it sadly i have to get a new one 😔
